File: Python3/1022.py
n=int(input())
title = {}
author = {}
key_words = {}
publisher = {}
pub_year = {}
for _ in range(n):
    id = int(input())
    name = input()
    if title.get(name) is None:
        title[name] = []
    title[name].append(id)
    author_name = input()
    if author.get(author_name) is None:
        author[author_name] = []
    author[author_name].append(id)
    words = input().split()
    for word in words:
        if key_words.get(word) is None:
            key_words[word] = []
        key_words[word].append(id)
    pub = input()
    if publisher.get(pub) is None:
        publisher[pub]=[]
    publisher[pub].append(id)
    year = input()
    if pub_year.get(year) is None:
        pub_year[year] = []
    pub_year[year].append(id)
m = int(input())
import re
def query(dic,string):
    try:
        for each in sorted(dic[string]):
            print("{:07d}".format(each))
    except:
        print("Not Found")
for _ in range(m):
    a, b = re.split(r'[\,\:]+', input())
    b=b.strip()
    print("{}: {}".format(a,b))
    if a == '1':
        query(title,b)
    elif a == '2':
        query(author,b)
    elif a == '3':
        query(key_words,b)
    elif a== '4':
        query(publisher,b)
    else:
        query(pub_year,b)

File: c++/1022.cpp

//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/9/29.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <string.h>
#include <set>
using namespace std;


map<string,set<int>> title,author,key_word,publisher,year;


int n,m,id;

void query(map<string,set<int>>& data,string &str){
    if(data.find(str) != data.end()){
        for(auto each:data[str]) printf("%07d\n",each);
    }else
        printf("Not Found\n");
}

int main(){
    cin >> n;
    string a,b,c,d,e;
    for(int i=0;i<n;++i){
        scanf("%d\n",&id);
        getline(cin,a);
        title[a].insert(id);
        getline(cin,b);
        author[b].insert(id);
        while (cin >> c) {
            key_word[c].insert(id);
            char ch = getchar();
            if(ch=='\n') break;
        }
        getline(cin,d);
        publisher[d].insert(id);
        getline(cin,e);
        year[e].insert(id);
    }
    scanf("%d",&m);
    for(int i=0;i<m;++i){
        int category;
        scanf("%d: ",&category);
        getline(cin,a);
        cout << category << ": "<<a <<"\n";
        if(category == 1) query(title, a);
        else if (category == 2)query(author, a);
        else if(category == 3) query(key_word, a);
        else if(category==4)query(publisher, a);
        else query(year, a);
    }
}
